File: crates/frances-workflow/src/modules/lifecycle.rs
//! `frances:v1/lifecycle` — graceful-shutdown hook.
//!
//! Exposes the user-facing `lifecycle` object (`{ shutdown: null }`).
//! Workflows opt in by assigning `lifecycle.shutdown = async () => {...}`.
//!
//! The module body (`js/lifecycle.js`) registers a *shutdown runner* on a
//! hidden global. The runtime invokes that runner when shutdown is
//! requested (a dehydrate from the host, or `workflow.exit()`), then
//! closes the inbox itself — it owns the `closed` / `closed_notify`
//! handles, so closing is a host concern, not the module's.

use rquickjs::{Ctx, Object, Result as JsResult};

/// Build the user-facing `lifecycle` object. Workflows assign
/// `lifecycle.shutdown`; the runtime reads it back (via the runner the
/// module registers) when winding the workflow down.
pub(crate) fn build_lifecycle_object<'js>(ctx: &Ctx<'js>) -> JsResult<Object<'js>> {
    let lifecycle = Object::new(ctx.clone())?;
    lifecycle.set("shutdown", rquickjs::Value::new_null(ctx.clone()))?;
    Ok(lifecycle)
}

File: crates/frances-workflow/src/modules/js/lifecycle.js
// `frances:v1/lifecycle` — graceful-shutdown hook.
//
// Workflows opt in by assigning a function:
//
//   import { lifecycle } from "frances:v1/lifecycle";
//   lifecycle.shutdown = async () => { /* save state, emit, ... */ };
//
// When the host requests shutdown (a dehydrate, or `workflow.exit()`),
// the runtime reads `lifecycle.shutdown` off this object, runs it, then
// closes the inbox so any `for await (const input of inbox)` loop in user
// code unwinds. The module just needs to export the object.

const __s = globalThis.__frances_v1_stash__;

export const lifecycle = __s.lifecycle;
